// type-graphql-v2-fork/dist/helpers/resolver-metadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getResolverMetadata = void 0;
const findType_1 = require("./findType");
const errors_1 = require("../errors");
function getResolverMetadata(prototype, propertyKey, returnTypeFunc, options = {}) {
    var _a;
    if (typeof propertyKey === 'symbol') {
        throw new errors_1.SymbolKeysNotSupportedError();
    }
    const { getType, typeOptions } = (0, findType_1.findType)({
        metadataKey: 'design:returntype',
        prototype,
        propertyKey,
        returnTypeFunc,
        typeOptions: options
    });
    const methodName = propertyKey;
    return {
        methodName,
        schemaName: (_a = options.name) !== null && _a !== void 0 ? _a : methodName,
        target: prototype.constructor,
        getReturnType: getType,
        returnTypeOptions: typeOptions,
        description: options.description,
        deprecationReason: options.deprecationReason,
        complexity: options.complexity
    };
}
exports.getResolverMetadata = getResolverMetadata;
